fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -179,69 +179,126 @@ private IndexingByIndex getIndexerByIndex() {
                     Object sourceIndexSubspaceKey = decodeSubspaceKey(conflictingIndexingTypeStamp.getSourceIndexSubspaceKey());
+                    IndexingPolicy origPolicy = indexingPolicy;
                     indexingPolicy = IndexingPolicy.newBuilder()
                             .setSourceIndexSubspaceKey(sourceIndexSubspaceKey)
-                            .forbidRecordScan() // no third chance
+                            .setForbidRecordScan(origPolicy.forbidRecordScan)
                             .build();
-                    return fallbackFunction.get();
+                    return indexingLauncher(indexingFunc, attemptCount, origPolicy);
                 }
-                // None of the above? let it fall through
+                // No other methods (yet). This line should never be reached.
             }
 
-            if (common.getIndexStatePrecondition() == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_POLICY_CHANGED) {
+            if (indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_POLICY_CHANGED) {
+                // Just rebuild
                 common.setIndexStatePrecondition(IndexStatePrecondition.BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY);
-                indexer = null;
-                return fallbackFunction.get();
+                return indexingLauncher(indexingFunc, attemptCount);
             }
 
-            if (common.getIndexStatePrecondition() == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_ERROR_IF_POLICY_CHANGED) {
-                throw FDBExceptions.wrapException(ex); // error it is
+            if (indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_ERROR_IF_POLICY_CHANGED) {
+                // Error it is
+                throw FDBExceptions.wrapException(ex);
             }
+
+            // All other precondition will be handled below
         }
-        if (indexingPolicy.isByIndex() &&
-                ! indexingPolicy.isForbidRecordScan() &&
-                ! fallbackToRecordsScan &&
-                (ex.getCause() instanceof IndexingByIndex.ValidationException)) {
-            // fallback to a records scan
-            if (LOGGER.isWarnEnabled()) {
-                final KeyValueLogMessage message =
-                        KeyValueLogMessage.build("fallback to records scan",
-                                LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
-                                LogMessageKeys.INDEXER_ID, common.getUuid(),
-                                LogMessageKeys.INDEX_NAME, index.getName());
-                LOGGER.warn(message.toString(), ex);
+
+        if (indexingPolicy.isByIndex() && IndexingByIndex.isValidationException(ex)) {
+            // Validation failed - the source index cannot be used for records scanning
+            if (requestedPolicy != null) {
```

Comment:
There's something about this that makes it seem like this should only be called if the user used
`BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_POLICY_CHANGED`, insofar as the index changing could arguably be seen as similar to a policy changing.... On the other hand, I think it makes sense for the default behavior to rebuild if the index changes in a way that makes the validation now break. So, I'm not certain.

Perhaps if the `IndexingPolicy` was more expressive (as I think we plan on doing) it would be an option on the build. On the other hand, I'm not sure we want to make an already fairly complicated set of options more complicated.

Reply:
I had the same dilemma. My rule of thumb was something like "never rebuild an index unless the caller had explicitly mentioned '_REBUILD_' in his request", but after testing the code in an existing project it seems to lead to a dead-end - as the existing project retry indexing with the same default parameters again and again as the indexing process fails again and again because the previous source-index was modified (or became unreadable). 
Since we can't educate every caller to handle this exception properly, and since a rebuild - when a previous source-index is modified - is unavoidable, doing it "under the hood" seems to be practical and reasonable. I hope that the performance gain of the by-index scan would make the penalty of this occasional rebuilding seem marginal.